src/providers/Web3Provider.tsx

```diff
@@ -60,35 +60,96 @@ export const Web3ContextProvider: FC<PropsWithChildren> = ({ children }) => {
     ...web3ProviderInitialState,
   })
 
+  const _connectionChanged = (isConnected: boolean) => {
+    setState((prevState) => ({
+      ...prevState,
+      isConnected,
+    }))
+  }
+
+  const _accountsChanged = (accounts: string[]) => {
+    if (accounts.length <= 0) {
+      _connectionChanged(false)
+      return
+    }
+
+    const [account] = accounts
+    setState(prevState => ({
+      ...prevState,
+      account,
+    }))
+  }
+
+  const _setNetworkSpecificVars = (chainId: number, sapphireEthProvider = state.sapphireEthProvider!): void => {
+    if (!sapphireEthProvider) {
+      throw new Error('[Web3Context] Sapphire provider is required!')
+    }
+
+    if (!(chainId in NETWORKS)) {
+      throw new UnknownNetworkError('Unknown network!')
+    }
+
+    const { wRoseContractAddress, explorerBaseUrl, networkName } = NETWORKS[chainId]
+
+    const wRoseContract = new ethers.Contract(
+      wRoseContractAddress,
+      WrappedRoseMetadata.output.abi,
+      sapphireEthProvider.getSigner(),
+    )
+
+    setState(prevState => ({
+      ...prevState,
+      wRoseContract,
+      explorerBaseUrl,
+      networkName,
+      wRoseContractAddress,
+    }))
+  }
+
+  const _chainChanged = () => {
+    window.location.reload()
+  }
+
+  const _connect = () => _connectionChanged(true)
+  const _disconnect = () => _connectionChanged(false)
+
+  const _addEventListeners = (ethProvider = window.ethereum) => {
+    ethProvider.on('accountsChanged', _accountsChanged)
+    ethProvider.on('chainChanged', _chainChanged)
+    ethProvider.on('connect', _connect)
+    ethProvider.on('disconnect', _disconnect)
+  }
+
+  const _removeEventListeners = (ethProvider = window.ethereum) => {
+    ethProvider.off('accountsChanged', _accountsChanged)
+    ethProvider.off('chainChanged', _chainChanged)
+    ethProvider.off('connect', _connect)
+    ethProvider.off('disconnect', _disconnect)
```

Comment:
This doesn't work, referenced functions aren't the old instance:
- disconnect all accounts from wrose
- add console.log to _accountsChanged and _init
- connect an account + change account
- see 1 '_accountsChanged' printed
- disconnect all accounts from wrose
- see 1 '_init' printed
- connect an account + change account
- see 2 '_accountsChanged' printed
- disconnect all accounts from wrose
- see 1 '_init' printed
- connect an account + change account
- see 3 '_accountsChanged' printed

Reply:
True, thanks for catching this 👍